fairlearn/reductions/_exponentiated_gradient/exponentiated_gradient.py

```diff
@@ -199,13 +204,21 @@ def predict(self, X):
     def _pmf_predict(self, X):
         """Probability mass function for the given input data.
 
+        Calculates predictions for each of the underlying classifiers. The predictions are
+        subsequently multiplied by the classifiers' weights and aggregated into the returned
+        probabilities.
+
         :param X: Feature data
         :type X: numpy.ndarray or pandas.DataFrame
         :return: Array of tuples with the probabilities of predicting 0 and 1.
         :rtype: pandas.DataFrame
         """
-        pred = pd.DataFrame()
-        for t in range(len(self._hs)):
-            pred[t] = self._hs[t](X)
+        n_samples = len(X)
+
+        # ignore classifiers with weight 0 since they can only contribute 0 to the result
+        pred = pd.DataFrame(0, index=range(n_samples), columns=range(len(self._hs))) \
+            .apply(lambda row: np.zeros(n_samples) if self._weights[int(row.name)] == 0
```

Comment:
Is this saving us a call to the predictors, or have those already happened? This statement is a bit of a monstrosity to set things to zero which would be zero anyway. Unless it saves an expensive call.

Reply:
The first line doesn't do anything other than create a data structure with zeros.
I would call it expensive if we have a billion rows and a thousand classifiers of which only 5 have non-zero weights :-)

Reply:
Fair enough, but I still think this is a monstrosity of a statement. Rather than a lambda, could you put an inline `def evaluate_for_nonzero_weight()` (on the line immediately preceding this one) and have that in the `apply()` method?

I grant that at least part of this is due to my lack of familiarity with pandas, but I do think that when a lambda is covering multiple lines, it should probably get a proper function definition. I also while I have used it in list comprehensions, I'm not a huge fan of the "stmt if cond else stmt" construct.

Reply:
Sure we can do that. The lambda is not the important part, the `apply` is, so we can certainly pre-define it.

Reply:
Are you going to?